Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp

```diff
@@ -55,161 +55,81 @@ FileSystemSyncAccessHandle::FileSystemSyncAccessHandle(ScriptExecutionContext& c
 {
-    ASSERT(isClosingOrClosed());
-
     m_source->unregisterSyncAccessHandle(m_identifier);
-
-    if (m_closeResult)
-        return;
-
-    // Task replies may be scheduled to WorkerRunLoop but do not run.
-    auto pendingPromises = std::exchange(m_pendingPromises, { });
-    for (auto& promise : pendingPromises) {
-        std::visit([](auto& promise) {
-            promise.reject(Exception { UnknownError, "AccessHandle is about to be destroyed"_s });
-        }, promise);
-    }
-
-    closeBackend(CloseMode::Sync);
+    closeInternal(ShouldNotifyBackend::Yes);
 }
 
-bool FileSystemSyncAccessHandle::isClosingOrClosed() const
+ExceptionOr<void> FileSystemSyncAccessHandle::truncate(unsigned long long size)
 {
-    return m_closeResult || !m_closeCallbacks.isEmpty();
-}
-
-void FileSystemSyncAccessHandle::truncate(unsigned long long size, DOMPromiseDeferred<void>&& promise)
-{
-    if (isClosingOrClosed())
-        return promise.reject(Exception { InvalidStateError, "AccessHandle is closing or closed"_s });
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
     auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
     if (!scope)
-        return promise.reject(Exception { InvalidStateError, "Context is invalid"_s });
-
-    m_pendingPromises.append(WTFMove(promise));
-    WorkerGlobalScope::postFileSystemStorageTask([weakThis = WeakPtr { *this }, file = m_file.handle(), size, workerThread = Ref { scope->thread() }]() mutable {
-        workerThread->runLoop().postTask([weakThis = WTFMove(weakThis), success = FileSystem::truncateFile(file, size)](auto&) mutable {
-            if (weakThis)
-                weakThis->completePromise(success ? ExceptionOr<void> { } : Exception { UnknownError });
-        });
-    });
-}
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-void FileSystemSyncAccessHandle::getSize(DOMPromiseDeferred<IDLUnsignedLongLong>&& promise)
-{
-    if (isClosingOrClosed())
-        return promise.reject(Exception { InvalidStateError, "AccessHandle is closing or closed"_s });
-
-    auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
-    if (!scope)
-        return promise.reject(Exception { InvalidStateError, "Context is invalid"_s });
-
-    m_pendingPromises.append(WTFMove(promise));
-    WorkerGlobalScope::postFileSystemStorageTask([weakThis = WeakPtr { *this }, file = m_file.handle(), workerThread = Ref { scope->thread() }]() mutable {
-        workerThread->runLoop().postTask([weakThis = WTFMove(weakThis), success = FileSystem::fileSize(file)](auto&) mutable {
-            if (weakThis)
-                weakThis->completePromise(success ? ExceptionOr<uint64_t> { success.value() } : Exception { UnknownError });
-        });
-    });
+    bool succeeded = FileSystem::truncateFile(m_file.handle(), size);
+    return succeeded ? ExceptionOr<void> { } : Exception { UnknownError };
```

Comment:
https://fs.spec.whatwg.org/#api-filesystemsyncaccesshandle-truncate is using InvalidStateError, maybe we should use this one with a message.
This may apply to other methods.

Reply:
Sounds good; will update.